schematic/base.py

```diff
@@ -0,0 +1,446 @@
+# -*- coding: utf-8 -*-
+# MIT License
+#
+# Copyright (c) 2019 Cody J. Hanson
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+import click
+from psycopg2 import sql
+from abc import ABC
+from csv import DictReader
+from queue import Queue
+from schematic import NameSqlMixin, DictableMixin
+
+
+class ColumnTypeNotFoundError(Exception):
+    pass
+
+
+class TableColumnType(ABC, NameSqlMixin, DictableMixin):
+    """Represents a type for a table column.
+
+    The "size" of two different column types can be compared using
+    the built in comparison operators. For two TableColumnTypes, A and B,
+    if A is a superset of B, then A > B evaluates to True.
+
+    Note that TableColumnTypes are only fully comparable when in the
+    same subtree. If A is not found when traversing next_less_restrictive
+    in B and vice versa, A < B, A > B, and A == B will all return False
+
+    Attributes:
+      name: The name of the type
+      next_less_restrictive: The TableColumnType which is next less restrictive
+      parameter: Instance parameter for this class. E.g., '256' for a VARCHAR(256)
+
+    TODO(Cody): Comparisons of incomparable operands should raise an exception
+    TODO(Cody): Factor out the "most to least restrictive" iterator logic into a method on this class.
+    TODO(Cody): Change the "parameter" attribute to use **kwargs.
+    """
+    name = "TableColumnType"
```

Comment:
```suggestion
    @property
    @abstractmethod
    def name(self):
        pass
```
force subclasses to pick their own name